packages/worker/src/worker.ts

```diff
@@ -993,260 +999,246 @@ export class Worker {
+          await this.processSinkCalls(calls, isReplaying);
+        }
+      }
+    } catch (error) {
+      if (error instanceof UnexpectedError) {
+        // rethrow and fail the worker
+        throw error;
+      }
+      this.log.error('Failed to activate workflow', {
+        ...(workflow ? workflow.logAttributes : { runId: activation.runId }),
+        error,
+        workflowExists: workflow !== undefined,
+      });
+      const completion = coresdk.workflow_completion.WorkflowActivationCompletion.encodeDelimited({
+        runId: activation.runId,
+        failed: {
+          failure: await encodeErrorToFailure(this.options.loadedDataConverter, error),
+        },
+      }).finish();
+      // We do not dispose of the Workflow yet, wait to be evicted from Core.
+      // This is done to simplify the Workflow lifecycle so Core is the sole driver.
+      return { state: undefined, output: { close: true, completion } };
+    }
+  }
+
+  protected async createWorkflow(
+    activation: coresdk.workflow_activation.WorkflowActivation,
+    startWorkflow: coresdk.workflow_activation.IStartWorkflow
+  ): Promise<WorkflowWithLogAttributes> {
+    const workflowCreator = this.workflowCreator!;
+    if (
+      !(
+        startWorkflow.workflowId != null &&
+        startWorkflow.workflowType != null &&
+        startWorkflow.randomnessSeed != null &&
+        startWorkflow.firstExecutionRunId != null &&
+        startWorkflow.attempt != null &&
+        startWorkflow.startTime != null
+      )
+    ) {
+      throw new TypeError(`Malformed StartWorkflow activation: ${JSON.stringify(startWorkflow)}`);
+    }
+    if (activation.timestamp == null) {
+      throw new TypeError('Got activation with no timestamp, cannot create a new Workflow instance');
+    }
+    const {
+      workflowId,
+      randomnessSeed,
+      workflowType,
+      parentWorkflowInfo,
+      workflowExecutionTimeout,
+      workflowRunTimeout,
+      workflowTaskTimeout,
+      continuedFromExecutionRunId,
+      continuedFailure,
+      lastCompletionResult,
+      firstExecutionRunId,
+      retryPolicy,
+      attempt,
+      cronSchedule,
+      workflowExecutionExpirationTime,
+      cronScheduleToScheduleInterval,
+      memo,
+      searchAttributes,
+    } = startWorkflow;
+
+    const workflowInfo: WorkflowInfo = {
+      workflowId,
+      runId: activation.runId,
+      workflowType,
+      searchAttributes:
+        (mapFromPayloads(searchAttributePayloadConverter, searchAttributes?.indexedFields) as SearchAttributes) || {},
+      memo: await decodeMapFromPayloads(this.options.loadedDataConverter, memo?.fields),
+      parent: convertToParentWorkflowType(parentWorkflowInfo),
+      lastResult: await decodeFromPayloadsAtIndex(this.options.loadedDataConverter, 0, lastCompletionResult?.payloads),
+      lastFailure: await decodeOptionalFailureToOptionalError(this.options.loadedDataConverter, continuedFailure),
```

Comment:
nit: if you're refactoring this, you can use `Promise.all` to run the codecs concurrently.

Reply:
Very good point, though that might also be addressed in a distinct PR. There are a few other places with the same issue. Should be easy to fix all of these in a single PR.

Reply:
Good point and I also vote let's do it in a follow-on PR. Although it's not a big thing, this PR is already touching a lot of critical and somewhat complex implementation, involving a library that we are slightly uncomfortable with in this context. So, in the event that we do discover some unexpected behavior, I would like us to have as few lines of investigation as possible.